clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -1512,6 +1512,57 @@ class CIRFuncLowering : public mlir::OpConversionPattern<mlir::cir::FuncOp> {
     }
   }
 
+  /// Do necessary lowering of the function attributes. The argument `attrs`
+  /// is used as input and output.
+  void lowerFuncAttributes(const mlir::TypeConverter *converter,
+                           SmallVectorImpl<mlir::NamedAttribute> &attrs) const {
+    mlir::AttrTypeReplacer replacer;
+    replacer.addReplacement(
+        [&](mlir::cir::OpenCLKernelMetadataAttr attr)
+            -> std::pair<mlir::Attribute, mlir::WalkResult> {
+          mlir::TypeAttr vecTypeHint = attr.getVecTypeHint();
+          // Lower the type in vector type hint from CIR to MLIR type,
+          // then we can use it directly in the module translation.
+          if (vecTypeHint) {
+            // See also clang::CodeGen::CodeGenFunction::EmitKernelMetadata
+            auto hintQTy = vecTypeHint.getValue();
+            auto hintEltQTy = dyn_cast<mlir::cir::VectorType>(hintQTy);
+            auto isCIRSignedIntType = [](mlir::Type t) {
+              return isa<mlir::cir::IntType>(t) &&
+                     cast<mlir::cir::IntType>(t).isSigned();
+            };
+            bool isSignedInteger =
+                isCIRSignedIntType(hintQTy) ||
+                (hintEltQTy && isCIRSignedIntType(hintEltQTy.getEltType()));
+            // Restore the signedness on integer types for future translation.
+            // OpenCL / SPIR-V requires the signedness of integer types, but
+            // we usually drops it in the type converter.
+            // The non-signless type resides in the attribute, which is a CIR
+            // attribute. It won't affect the verification of LLVM dialect.
+            auto signedness = isSignedInteger ? mlir::IntegerType::Signed
+                                              : mlir::IntegerType::Unsigned;
+            auto loweredHintQTy = converter->convertType(hintQTy).replace(
```

Comment:
You don't need all this stuff, just update the attribute with the new type with the result of `converter->convertType(..)`, you could use the auto generated methods for setting `OpenCLKernelMetadataAttr` members. If updating doesn't work (I don't why it shouldn't), create a new attribute at that point to replace. You shouldn't need to call into `addReplacement`.

You should also:
- Store an extra member for signedness in `OpenCLKernelMetadataAttr` and add a comment in the description that the type could be either CIR or LLVM, depending of the lowering level. It doesn't matter CIR already encode that info - logic is just too complicated otherwise and not worth the trouble.
- Add an extraDeclaration `isSignedHint` or similar that checks on that member, so the logic can be abstracted from here and only used in `LowerToLLVMIR.cpp`. `OpenCLKernelMetadataAttr`'s verifier should make sure that if the type is CIR, the sign should match what that member encodes.
- `typeConverter` should already be available, see other examples around in this file.